examples/state_machine/states.py

```diff
@@ -0,0 +1,517 @@
+import events
+import effects
+
+from effects import PNEffect
+from typing import Union
+
+from pubnub.enums import PNOperationType, PNStatusCategory
+from pubnub.exceptions import PubNubException
+
+
+class PNContext(dict):
+    channels: list
+    groups: list
+    tt: int
+    handshake_count: int
+    error: PubNubException
+
+    def update(self, context):
+        super().update(context.__dict__)
+
+
+class PNStatus:
+    category: PNStatusCategory
+    operation: PNOperationType
+
+    def __init__(self, category: Union[None, PNStatusCategory], operation: Union[None, PNOperationType]) -> None:
+        self.category = category
+        self.operation = operation
+
+
+class PNState:
+    _context: PNContext
+
+    def __init__(self, context: PNContext) -> None:
+        self._context = context
+        self._transitions = {}
+
+    def on(self, event: events.PNEvent, context: PNContext) -> PNEffect:
+        return self._transitions[event.get_name()](event, context)
+
+    def on_enter(self, context: Union[None, PNContext]):
+        pass
+
+    def on_exit(self):
+        pass
+
+    def get_context(self) -> PNContext:
+        return self._context
+
+
+class PNTransition:
+    context: PNContext
+    state: PNState
+    effect: Union[None, list[PNEffect]]
+    status: Union[None, list[PNStatus]]
```

Comment:
Why? EmitStatus is just another effect, so I don't think this `status` here is necessary